### ISSUE TITLE
Pro Oracle oracleTSTZ doesn't unwrap connection from pool prior to TZ constructor

### ISSUE THREAD

**Issue by Nthalk** [jOOQ/jOOQ#9088]
### Expected behavior and actual behavior:

When using `<javaTimeTypes>` and a `connection pool` with an `ojdbc6` data source, I expect to be able a query similar to:

`db.selectFrom(RET_JOB).where(RET_JOB.CREATED_AT.greaterThan(OffsetDateTime.now().minusHours(1)).fetch()`

To return a list of records.

What I get is a class cast exception (`java.lang.ClassCastException: org.apache.tomcat.dbcp.dbcp.PoolingDataSource$PoolGuardConnectionWrapper cannot be cast to oracle.jdbc.OracleConnection`), since my connection is a pooled connection and the pro function: `oracleTSTZ` is not unwrapping the `connection` with `targetLocalConnection()`:
```
private static final Object oracleTSTZ(Connection connection, OffsetDateTime val) {
            Calendar cal = Calendar.getInstance();
            cal.setTimeZone(TimeZone.getTimeZone(val.getOffset()));
            cal.setTimeInMillis(val.toInstant().toEpochMilli());

            // [#7400] Avoid the newer TIMESTAMPTZ(Connection, Timestamp, ZoneId) constructor,
            // which is available only on ojdbc8. Clients may use an outdated driver, despite
            // using Java 8+
            return Reflect.on("oracle.sql.TIMESTAMPTZ")
                          .create(connection, Timestamp.from(val.toInstant()), cal)
                          .get();
        }
```

When I run in debug mode, and force-override `connection` with `connection = localTargetConnection()` then it works as expected.

It seems like this is fixed for `createOracleARRAY`:

```
 private static final Array createOracleARRAY(Configuration configuration, ArrayRecord<?> array) {

            // [#1544] We can safely assume that localConfiguration has been
            // set on DefaultBindContext, prior to serialising arrays to SQLOutput
            Array result = on(localTargetConnection()).call("createARRAY", Tools.getMappedArrayNameString(configuration, array), unnest(array)).<Array>get();
            DefaultExecuteContext.register(result);
            return result;
        }
```

### Versions:

- jOOQ: 3.11.12 (pro)
- Java: 1.8
- Database (include vendor): Oracle 12
- OS: osx db client, linux db server
- JDBC Driver (include name if inofficial driver): ojdbc6 11.2.0.4

**Comment by Nthalk** [comment]
Rewriting the bytecode for `DefaultOffsetDateTimeBinding` to  call `ctx.statement().setObject(ctx.index(), oracleTSTZ(localTargetConnection(), value));` does fix the issue, however it would be nice if this was in main line.

**Comment by knutwannheden** [comment]
Thank you very much for the detailed report and analysis. This indeed looks like a bug.

**Comment by knutwannheden** [comment]
The bug has been fixed for the 3.12 release as suggested.

**Comment by lukaseder** [comment]
I agree with the fix. Specifically, as this is implicitly using the new `UnwrapperProvider` SPI, which allows users to override the JDBC `Wrapper::unwrap` calls: #8145.

We'll backport this fix to 3.11.13 (#9095).
